File: trip-kit/src/main/java/skedgo/tripkit/routing/a2b/A2bRoutingRequest.java
package skedgo.tripkit.routing.a2b;

import org.immutables.value.Value;
import org.immutables.value.Value.Style.ImplementationVisibility;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import kotlin.Pair;

@Value.Immutable
@Value.Style(visibility = ImplementationVisibility.PACKAGE)
public abstract class A2bRoutingRequest {
  @NotNull public static Builder builder() {
    return null;
  }

  public abstract @NotNull Pair<Double, Double> origin();
  public abstract @NotNull Pair<Double, Double> destination();
  public abstract @Nullable String originAddress();
  public abstract @Nullable String destinationAddress();

  public interface Builder {
    Builder origin(@NotNull Pair<Double, Double> origin);
    Builder destination(@NotNull Pair<Double, Double> destination);
    Builder originAddress(@Nullable String originAddress);
    Builder destinationAddress(@Nullable String destinationAddress);
    A2bRoutingRequest build();
  }
}

File: TripKitDomain/src/main/java/com/skedgo/tripkit/a2brouting/A2bRoutingRequest.kt
package com.skedgo.tripkit.a2brouting

import org.immutables.value.Value
import org.immutables.value.Value.Style.ImplementationVisibility

@Value.Immutable
@Value.Style(visibility = ImplementationVisibility.PACKAGE)
abstract class A2bRoutingRequest {
  companion object {
    fun builder(): Builder = ImmutableA2bRoutingRequest.builder()
  }

  abstract val origin: Pair<Double, Double>
  abstract val destination: Pair<Double, Double>
  abstract val originAddress: String?
  abstract val destinationAddress: String?
  abstract val time: RequestTime

  interface Builder {
    fun origin(origin: Pair<Double, Double>): Builder
    fun destination(destination: Pair<Double, Double>): Builder
    fun originAddress(originAddress: String?): Builder
    fun destinationAddress(destinationAddress: String?): Builder
    fun time(time: RequestTime): Builder
    fun build(): A2bRoutingRequest
  }
}
